### PR TITLE
MSI-874: Api-Functional Tests For MSI are Failed.

### PR DESCRIPTION
**Pull request by nmalevanec**
### Description
Fix failed MSI REST and SOAP tests.

### Fixed Issues (if relevant)
1. magento-engcom/msi#874: Api-Functional Tests For MSI are Failed.

### Manual testing scenarios
1. Install Magento with MSI.
2. Configure dev/test/api-functional/phpunit_rest.xml and dev/test/api-functional/phpunit_soap.xml
3. Run api-function with REST adapter ex:
```
#!/bin/bash

RED='\e[1;31m'
GREEN='\e[1;32m'
BLUE='\e[1;34m'
ORANGE='\e[1;33m'
NC='\e[0m' # No Color

echo -e "${BLUE}Running MSI testsuite...${NC}";
php $PWD/vendor/phpunit/phpunit/phpunit --configuration $PWD/dev/tests/api-functional/phpunit_rest.xml --testsuite "msi"
```
4. Check all test are green. No failures or errors occured.
5. Run api-functional tests with SOAP adapter. ex:
```
#!/bin/bash

RED='\e[1;31m'
GREEN='\e[1;32m'
BLUE='\e[1;34m'
ORANGE='\e[1;33m'
NC='\e[0m' # No Color

echo -e "${BLUE}Running MSI testsuite...${NC}";
php $PWD/vendor/phpunit/phpunit/phpunit --configuration $PWD/dev/tests/api-functional/phpunit_soap.xml --testsuite "msi"
```
6. Check no errors and failures occured. Some tests may be skipped.

### CODE REVIEW

File: app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php [hunk 1]
```diff
@@ -77,24 +91,63 @@ public function dataProviderWrongParameters(): array
...
+                    ],
                 ],
             ],
         ];
     }
+
+    /**
+     * @inheritdoc
+     */
+    protected function _checkWrappedErrors($expectedWrappedErrors, $errorDetails)
```
Comment (maghamed) [inline app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php:131]: why to add protected method with name starting from underscore to the newly created code?

File: app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php [hunk 2]
```diff
@@ -77,24 +91,63 @@ public function dataProviderWrongParameters(): array
...
+            if (is_object($error)) {
+                $actualErrors[] = $error->message;
+            } else {
+                $actualErrors[] = $error;
+            }
+        }
+        $this->assertEquals(
+            $expectedErrors,
+            $actualErrors,
+            "Wrapped errors in fault details are invalid."
```
Comment (maghamed) [inline app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php]: single quotes should be used instead

File: app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/DeleteSourceItemConfigurationTest.php [hunk 3]
```diff
@@ -14,13 +14,14 @@
...
-    const SERVICE_NAME_DELETE = 'InventoryLowQuantityNotificationApiDeleteSourceItemConfigurationV1';
-    const SERVICE_NAME_GET = 'InventoryLowQuantityNotificationApiGetSourceItemConfigurationV1';
+    const SERVICE_NAME_DELETE = 'inventoryLowQuantityNotificationApiDeleteSourceItemConfigurationV1';
+    const SERVICE_NAME_GET = 'inventoryLowQuantityNotificationApiGetSourceItemConfigurationV1';
 
     /**
      * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/sources.php
      * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source_items.php
      * @magentoDataFixture ../../../../app/code/Magento/InventoryLowQuantityNotificationApi/Test/_files/source_item_configuration.php
+     * @magentoDbIsolation disabled
```
Comment (maghamed) [inline app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/DeleteSourceItemConfigurationTest.php]: why do we explicitly disable DB Isolation here?

File: app/code/Magento/InventoryBundle/Test/Api/BundleProductShouldBeInStockOnCreateTest.php [hunk 4]
```diff
@@ -81,9 +82,7 @@ public function testIsBundleProductWithSimpleProductInStockAfterCreate()
             ],
             "extension_attributes" => [
                 "bundle_product_options" => $bundleProductOptions,
-                'stock_item' => [
-                    'is_in_stock' => 1,
-                ],
+                'stock_item' => $this->getStockItemData()
```
Comment (maghamed) [inline app/code/Magento/InventoryBundle/Test/Api/BundleProductShouldBeInStockOnCreateTest.php:85]: why do we need to specify all the StockItem data? why does fallback not work?

File: app/code/Magento/InventoryApi/Test/Api/SourceRepository/CarrierLinkManagementTest.php [hunk 5]
```diff
@@ -165,8 +165,29 @@ public function testCarrierLinksValidation(array $carrierData, array $expectedEr
             $this->_webApiCall($serviceInfo, ['source' => $carrierData]);
             $this->fail('Expected throwing exception');
         } catch (\Exception $e) {
-            self::assertEquals($expectedErrorData, $this->processRestExceptionResult($e));
-            self::assertEquals(\Magento\Framework\Webapi\Exception::HTTP_BAD_REQUEST, $e->getCode());
+            if (TESTS_WEB_API_ADAPTER == self::ADAPTER_REST) {
+                self::assertEquals($expectedErrorData, $this->processRestExceptionResult($e));
+                self::assertEquals(\Magento\Framework\Webapi\Exception::HTTP_BAD_REQUEST, $e->getCode());
+            } elseif (TESTS_WEB_API_ADAPTER == self::ADAPTER_SOAP) {
```
Comment (maghamed) [inline app/code/Magento/InventoryApi/Test/Api/SourceRepository/CarrierLinkManagementTest.php]: please always use strict comparison wherever possible ===

File: app/code/Magento/InventoryApi/Test/Api/SourceItemsSave/ValidationTest.php [hunk 6]
```diff
@@ -373,7 +383,14 @@ public function failedValidationRelatedOnlyForRestDataProvider(): array
     public function testCreateWithEmptyData()
     {
         $sourceItems = [];
-        $expectedErrorData = ['message' => 'Input data is empty'];
+        if (TESTS_WEB_API_ADAPTER == self::ADAPTER_SOAP) {
```
Comment (maghamed) [inline app/code/Magento/InventoryApi/Test/Api/SourceItemsSave/ValidationTest.php]: strict comparison should be used

File: app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php [hunk 7]
```diff
@@ -23,8 +23,8 @@ class GetSourceItemConfigurationTest extends WebapiAbstract
      */
     public function testGetSourceItemConfiguration()
     {
-        $sourceCode = 'eu-1';
-        $sku = 'SKU-1';
+        $sourceCode = 'eu-2';
+        $sku = 'SKU-3';
```
Comment (maghamed) [inline app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php]: what's the reason to change source code / sku in this case? 

as we have
```
    [
        SourceItemInterface::SOURCE_CODE => 'eu-1',
        SourceItemInterface::SKU => 'SKU-1',
        SourceItemInterface::QUANTITY => 5.5,
        SourceItemInterface::STATUS => SourceItemInterface::STATUS_IN_STOCK,
    ],
```
in our fixture data

File: app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php [hunk 8]
```diff
@@ -14,7 +14,7 @@
 class GetSourceItemConfigurationTest extends WebapiAbstract
 {
     const RESOURCE_PATH = '/V1/inventory/source-item-configuration';
-    const SERVICE_NAME = 'InventoryLowQuantityNotificationApiGetSourceItemConfigurationV1';
+    const SERVICE_NAME = 'inventoryLowQuantityNotificationApiGetSourceItemConfigurationV1';
```
Comment (maghamed) [inline app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php]: resource path should be modified as we don't have SourceItem configuration except LowStock one
```
    <route url="/V1/inventory/source-item-configuration/:sourceCode/:sku" method="GET">
        <service class="Magento\InventoryLowQuantityNotificationApi\Api\GetSourceItemConfigurationInterface" method="execute"/>
        <resources>
            <resource ref="Magento_InventoryApi::source"/>
        </resources>
    </route>
```

all the usages of `source-item-configuration` should be modified to Low Quantity Notification

File: app/code/Magento/InventoryIndexer/Plugin/InventoryApi/InvalidateAfterEnablingOrDisablingSourcePlugin.php [hunk 9]
```diff
@@ -55,7 +55,7 @@ public function aroundSave(
         SourceInterface $source
     ) {
         $invalidationRequired = $this->isInvalidationRequiredForSource->execute(
-            $source->getSourceCode(),
+            (string)$source->getSourceCode(),
```
Comment (maghamed) [inline app/code/Magento/InventoryIndexer/Plugin/InventoryApi/InvalidateAfterEnablingOrDisablingSourcePlugin.php]: why do you have null returned here?

**Review by maghamed — CHANGES_REQUESTED** [review]

File: app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php (hunk 1 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php:131]: @maghamed, as original _checkWrappedErrors():
https://github.com/magento-engcom/msi/blob/f2fdae5246c1b859e4ce7532458dd1243eff0991/dev/tests/api-functional/framework/Magento/TestFramework/TestCase/WebapiAbstract.php#L639
 can not handle ValidationException without detail node thrown by https://github.com/magento-engcom/msi/blob/f2fdae5246c1b859e4ce7532458dd1243eff0991/app/code/Magento/Inventory/Model/StockSourceLink/Command/StockSourceLinksSave.php#L68 I had to rewrite this paricular method with my own.

File: app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php (hunk 2 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryCatalog/Test/Api/StockSourceLink/PreventAssignSourcesToDefaultStockTest.php]: Fixed

File: app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/DeleteSourceItemConfigurationTest.php (hunk 3 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/DeleteSourceItemConfigurationTest.php]: @maghamed, removed

File: app/code/Magento/InventoryBundle/Test/Api/BundleProductShouldBeInStockOnCreateTest.php (hunk 4 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryBundle/Test/Api/BundleProductShouldBeInStockOnCreateTest.php:85]: @maghamed I order to pass SOAP validation, we have to specify all these fields  as we create bundle product with stock item on the fly. Looks like it's been always that way:
https://github.com/magento-engcom/msi/blob/f2fdae5246c1b859e4ce7532458dd1243eff0991/dev/tests/api-functional/testsuite/Magento/Catalog/Api/ProductRepositoryInterfaceTest.php#L340

File: app/code/Magento/InventoryApi/Test/Api/SourceRepository/CarrierLinkManagementTest.php (hunk 5 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryApi/Test/Api/SourceRepository/CarrierLinkManagementTest.php]: @maghamed, fixed

File: app/code/Magento/InventoryApi/Test/Api/SourceItemsSave/ValidationTest.php (hunk 6 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryApi/Test/Api/SourceItemsSave/ValidationTest.php]: @maghamed, fixed

File: app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php (hunk 7 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php]: @maghamed, replaced. But with qty=5.6. As it now in fixture.

File: app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php (hunk 8 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryLowQuantityNotificationApi/Test/Api/GetSourceItemConfigurationTest.php]: @maghamed, renamed to /V1/inventory/low-quantity-notification.

File: app/code/Magento/InventoryIndexer/Plugin/InventoryApi/InvalidateAfterEnablingOrDisablingSourcePlugin.php (hunk 9 above)
Comment (nmalevanec) [reply app/code/Magento/InventoryIndexer/Plugin/InventoryApi/InvalidateAfterEnablingOrDisablingSourcePlugin.php]: @maghamed, as we have validation test with "source code" = null and validation happens after this plugin, we get fatal error here.

Comment (naydav) [reply app/code/Magento/InventoryIndexer/Plugin/InventoryApi/InvalidateAfterEnablingOrDisablingSourcePlugin.php]: If `source_code` is empty we don't need to perform MySql query
Need to skip invalidation

Comment (nmalevanec) [reply app/code/Magento/InventoryIndexer/Plugin/InventoryApi/InvalidateAfterEnablingOrDisablingSourcePlugin.php]: @naydav Done

**Pull request merged**